internals/overlord/servstate/manager_test.go

```diff
@@ -1492,65 +1509,89 @@ func (s *S) TestReapZombies(c *C) {
 	// variable set so the "service" knows to create a zombie child.
 	testExecutable, err := os.Executable()
 	c.Assert(err, IsNil)
+	exitChildPath := filepath.Join(s.dir, "exit-child")
+
 	layer := parseLayer(c, 0, "layer", fmt.Sprintf(`
 services:
     test2:
         override: replace
         command: %s
         environment:
-            PEBBLE_TEST_CREATE_ZOMBIE: 1
+            PEBBLE_TEST_CREATE_CHILD: 1
+            PEBBLE_TEST_EXIT_CHILD: %s
         on-success: ignore
-`, testExecutable))
+`, testExecutable, exitChildPath))
 	err = s.manager.AppendLayer(layer)
 	c.Assert(err, IsNil)
 
 	s.startServices(c, []string{"test2"}, 1)
-	s.waitUntilService(c, "test2", func(svc *servstate.ServiceInfo) bool {
-		return svc.Current == servstate.StatusActive
-	})
 
-	// Get the PID of the zombie child (the createZombie process printed it out)
+	// The child process creates a grandchild and will print the
+	// PID of the grandchild for us to inspect here. We need to
+	// wait until we observe this on stdout.
 	childPidRe := regexp.MustCompile(`.* childPid (\d+)`)
-	matches := childPidRe.FindStringSubmatch(s.logBufferString())
-	c.Assert(len(matches), Equals, 2)
-	childPid, err := strconv.Atoi(matches[1])
-	c.Assert(err, IsNil)
-
-	// Wait till it becomes a zombie (by reading /proc/<pid>/stat)
-	var zombied bool
-	for i := 0; i < 10; i++ {
-		stat, err := ioutil.ReadFile(fmt.Sprintf("/proc/%d/stat", childPid))
-		c.Assert(err, IsNil)
-		statFields := strings.Fields(string(stat))
-		c.Assert(len(statFields) >= 3, Equals, true)
-		state := statFields[2]
-		if state == "Z" { // Z for Zombie!
-			zombied = true
-			break
+	childPid := 0
+	timeout := time.After(10 * time.Second)
+	ticker := time.NewTicker(time.Millisecond)
+pid:
+	for {
+		select {
+		case <-ticker.C:
+			// Get the log without resetting it on read
+			matches := childPidRe.FindStringSubmatch(s.readLogBuffer())
+			if len(matches) == 2 {
+				childPid, err = strconv.Atoi(matches[1])
+				if err == nil {
+					// Valid grandchild PID
+					break pid
+				}
+			}
+		case <-timeout:
+			c.Fatalf("timed out waiting for grandchild pid on stdout")
 		}
-		time.Sleep(10 * time.Millisecond)
 	}
-	if !zombied {
-		c.Fatalf("timed out waiting for zombie to be created")
+	s.clearLogBuffer()
+
+	// Wait until the grandchild is zombified
+	timeout = time.After(10 * time.Second)
+zombi:
+	for {
+		select {
+		case <-ticker.C:
+			stat, err := ioutil.ReadFile(fmt.Sprintf("/proc/%d/stat", childPid))
+			c.Assert(err, IsNil)
+			statFields := strings.Fields(string(stat))
+			c.Assert(len(statFields) >= 3, Equals, true)
+			if statFields[2] == "Z" {
+				break zombi
+			}
+
+		case <-timeout:
+			c.Fatalf("timed out waiting for grandchild to zombify")
+		}
 	}
 
-	// Wait till the main process terminates
+	// Wait till the child terminates (test2 exits)
+	fd, err := os.OpenFile(exitChildPath, os.O_RDONLY|os.O_CREATE, 0666)
+	c.Assert(err, IsNil)
+	fd.Close()
```

Comment:
I'm not sure what this OpenFile ... immediately Close sequence does -- can you clarity in the comment?

Reply:
I was trying to ```touch``` a file, so that it exists and allows the other process to proceed? Maybe there is a more compact way, but I could not find one :)

Reply:
Perhaps `os.Create`? Though it's not quite the same flags -- but I'm guessing that doesn't matter in this case.

But my main concern is just adding something like "so that it exists and allows the other process to proceed" to the comment. It's hard to connect the current comment with creating the file.